### PR TITLE
Elements with negative margins do not avoid floats when appropriate

### PR DESCRIPTION
**Pull request by Ahmad-S792**
#### c80dcb41f40298f9d1251e3773cda13fe5f3c910
<pre>
Elements with negative margins do not avoid floats when appropriate

Elements with negative margins do not avoid floats when appropriate
<a href="https://bugs.webkit.org/show_bug.cgi?id=247076">https://bugs.webkit.org/show_bug.cgi?id=247076</a>

Reviewed by Alan Baradlay.

Merge - <a href="https://src.chromium.org/viewvc/blink?view=revision&amp">https://src.chromium.org/viewvc/blink?view=revision&amp</a>;revision=180708

This is to align Webkit behavior with Blink / Chrome and Gecko / Firefox and web-specfication.

Web-Spec: <a href="https://www.w3.org/TR/CSS2/visuren.html#floats">https://www.w3.org/TR/CSS2/visuren.html#floats</a>

&quot;The border
box of a table, a block-level replaced element, or an element in the normal flow
that establishes a new block formatting context (such as an element with
&apos;overflow&apos; other than &apos;visible&apos;) must not overlap the margin box of any floats
in the same block formatting context as the element itself.&quot;

When an element is replaced or has an overflow clip it should avoid floats
even if a negative margin wants to pull it back onto the floats.
Simplify and correct the function that looks after this. We always want to
position the element after any floats so we always take the greater of the
border/padding + margin or the offset required to avoid floats. If there is
no offset required to avoid floats then it&apos;s ok to let the negative margin
pull the element back over the start edge of its containing block.

* Source/WebCore/rendering/RenderBlockFlow.cpp:
(RenderBlockFlow::marginOffForSelfCollapsingBlock): Following:
1) New variable &quot;initialStartPosition&quot; with &quot;startPosition&quot; value
2) Add new variable &quot;positionToAvoidFloats&quot;
3) Add logic for child with an offset from the content edge to avoid floats
* Source/WebCore/rendering/RenderBlock.cpp:
(RenderBlock::computeStartPositionDeltaForChildAvoidingFloats): Deleted
* Source/WebCore/rendering/RenderBlock.h: Remove &quot;computeStartPositionDeltaForChildAvoidingFloats&quot; reference
* LayoutTests/fast/block/float/avoid-floats-with-negative-margins.html: Added Test Case
* LayoutTests/fast/block/float/avoid-floats-with-negative-margins-expected.html: Added Test Case Expectations
* LayoutTests/platform/ios/fast/block/float/032-expected.txt: Updated Test Expectations
* LayoutTests/platform/mac/fast/block/float/032-expected.txt: Updated Test Expectations
* LayoutTests/platform/mac/fast/block/float/032-expected.png: Updated Test Expectations

Canonical link: <a href="https://commits.webkit.org/256132@main">https://commits.webkit.org/256132@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/d1ed086834c192ef283fa675a869b0bd92d0da29

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94792 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3944 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/27671 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/104398 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/164663 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/98788 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4018 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/32126 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/87095 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100319 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100459 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2895 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/81152 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29895 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84818 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/84365 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72779 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38535 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/18178 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/36372 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/19459 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4233 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40293 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/42110 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42269 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38690 "Passed tests") | |

### CODE REVIEW

File: Source/WebCore/rendering/RenderBlockFlow.cpp [hunk 1]
```diff
@@ -933,18 +933,28 @@ LayoutUnit RenderBlockFlow::marginOffsetForSelfCollapsingBlock()
...
+
+        // If the child has an offset from the content edge to avoid floats then use that, otherwise let any negative
+        // margin pull it back over the content edge or any positive margin push it out.
+        // If the child is being centred then the margin calculated to do that has factored in any offset required to
+        // avoid floats, so use it if necessary.
+
+        if (style().textAlign() == TextAlignMode::WebKitCenter || child.style().marginStartUsing(&style()).isAuto())
+            newPosition = std::max(newPosition, positionToAvoidFloats + childMarginStart);
+        else if (positionToAvoidFloats > initialStartPosition)
+            newPosition = std::max(newPosition, positionToAvoidFloats);
```
Comment (alanbaradlay) [inline Source/WebCore/rendering/RenderBlockFlow.cpp]: these lines look intended .

**Review by alanbaradlay — APPROVED** [review]
please fix the indentation before merging

File: Source/WebCore/rendering/RenderBlockFlow.cpp (hunk 1 above)
Comment (smfr) [reply Source/WebCore/rendering/RenderBlockFlow.cpp]: indented?

Comment (alanbaradlay) [reply Source/WebCore/rendering/RenderBlockFlow.cpp]: lol yes, indented (luckily, didn't mess it up the second time, down there)

Comment (Ahmad-S792) [reply Source/WebCore/rendering/RenderBlockFlow.cpp]: Good spot - fixed!!!

**Comment by Ahmad-S792** [comment]
> please fix the indentation before merging

Done in next push, which is now running via EWS. Once all major build passes, will merge.

**Pull request merged**